### PR TITLE
Setup Project

### PR DESCRIPTION
**Pull request by CMDRyan**
Initiate required files 

- gitlabAPI.js : Defined to fetch Projects and Jobs using API provided by gitlab.
- gitJobs.js : Get and filter required data from gitlab using provided group ID.
- app.js: Localhost server on port: 8081.
Additionally, empty ejs template engine views are added.

Usage:

- $export API_TOKEN=`your token`
- $node app.js

- Browser: http://localhost:8081/groups/ `group ID` /jobs

### CODE REVIEW

File: getJobs.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+module.exports.getJobs = function getJobs(requestedGroupID) {
+  return new Promise((resolve) => {
+
+    const apiToken = process.env.API_TOKEN; 
+    const gitlabAPI = require("./gitlabAPI");
```
Comment (Steven-Chan) [inline getJobs.js:5]: Use require at top level in this file.

File: getJobs.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+module.exports.getJobs = function getJobs(requestedGroupID) {
+  return new Promise((resolve) => {
+
+    const apiToken = process.env.API_TOKEN;
```
Comment (Steven-Chan) [inline getJobs.js:4]: Constants should also be put in top level, or you can create a new file to put all server config called `config.js` and put all env var in it.

File: getJobs.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+module.exports.getJobs = function getJobs(requestedGroupID) {
+  return new Promise((resolve) => {
```
Comment (Steven-Chan) [inline getJobs.js:2]: You don't need to create a new promise in this way. We only `new Promise` when transforming non-promise functions to promise.

File: getJobs.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    gitlabAPI
+      .getProjectsByGroupID(groupID, apiToken, getProjectSearchParams)
+
+      .then((data) => {
+        const ids = [];
+        for (let i = 0; i < data.length; i += 1) {
+          const item = data[i];
+          ids[i] = item.id;
+        }
+        return ids;
```
Comment (Steven-Chan) [inline getJobs.js:21]: Try using `map` of array to simplify the code.

File: getJobs.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      })
+      .then((projectIDs) => {
+        const data = [];
+        for (let i = 0; i < projectIDs.length; i += 1) {
+          data[i] = gitlabAPI.getJobsByProjectID(projectIDs[i], apiToken);
+        }
+        return Promise.all(data);
```
Comment (Steven-Chan) [inline getJobs.js:28]: Try `map`.

File: getJobs.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            jobStatus === "success" ||
+            jobStatus === "canceled" ||
+            jobStatus === "failed" ||
+            jobStatus === "skipped"
+          ) {
+          } else {
+            outputJobs[j] = sortedData[i];
+            j++;
+          }
+        }
```
Comment (Steven-Chan) [inline getJobs.js:49]: Try `filter`.

File: gitlabAPI.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
+const fetch = require("node-fetch");
+
+exports.getProjectsByGroupID = function getProjectsByGroupID(
+  GroupID,
```
Comment (Steven-Chan) [inline gitlabAPI.js:4]: Do not use capital letter to start a param variable name.

File: gitlabAPI.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      },
+    }
+  );
+  return projects.then((response) => {
+    const data = response.json();
+    return data;
+  });
+};
+
+exports.getJobsByProjectID = function getJobsByProjectID(ProjectID, key) {
```
Comment (Steven-Chan) [inline gitlabAPI.js:22]: Do not use capital letter to start a param variable name.

File: gitlabAPI.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+const fetch = require("node-fetch");
+
+exports.getProjectsByGroupID = function getProjectsByGroupID(
+  GroupID,
+  key,
+  searchParam
```
Comment (Steven-Chan) [inline gitlabAPI.js:6]: Please organize the function in better way. JS function param is unnamed, so the ordering is important. We group similar things together, so `key` should not be in the middle of `groupID` and `searchParams`.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    "@typescript-eslint/parser": "^4.28.3",
+    "autoprefixer": "^10.3.1",
+    "eslint-config-airbnb": "^18.2.1",
+    "eslint-config-prettier": "^8.3.0",
+    "eslint-plugin-import": "^2.23.4",
+    "eslint-plugin-jsx-a11y": "^6.4.1",
+    "eslint-plugin-react": "^7.24.0",
+    "eslint-plugin-react-hooks": "^4.2.0",
+    "postcss": "^8.3.5",
+    "prettier": "2.3.2"
```
Comment (Steven-Chan) [inline package.json:30]: Please pin all to exact version and run `yarn` again.

File: views/partials/head.ejs [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<meta charset="UTF-8">
+<title>Testing title</title>
+<!-- CSS (load bootstrap from a CDN) -->
+<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.5.2/css/bootstrap.min.css">
```
Comment (Steven-Chan) [inline views/partials/head.ejs:4]: We use tailwind instead of bootstrap, so please remove this.

File: views/src/styles.css [hunk 12]
```diff
@@ -0,0 +1,133 @@
+@import url("https://fonts.googleapis.com/css2?family=Quicksand&display=swap");
```
Comment (Steven-Chan) [inline views/src/styles.css:1]: This PR does not have UI yet, so please remove this file.

On the other hand, it looks to me that defining style here defeat the purpose of using tailwind.

**Review by Steven-Chan — CHANGES_REQUESTED** [review]
1. Please remove blackbox related thing in this commit
2. Do not check in tailwind.css here, it is a generated file.
3. Put source code files in `src` folder instead of the top level. Otherwise you will mix build phase code and runtime code together.

Put the steps to setup this project in README instead of PR comment.

**Comment by Steven-Chan** [comment]
testing

File: package.json (hunk 10 above)
Comment (Steven-Chan) [reply package.json:30]: Not addressed

**Comment by Steven-Chan** [comment]
@CMDRyan 

- blackbox not removed yet
- The tailwind.css is needed, but not committed and no instruction about it
- Dont say discussed minor issue in commit, list out what u have changed
- It is impossible to have both package-lock.json and yarn-lock.json at the same time

**Comment by CMDRyan** [comment]
@Steven-Chan 

- Resent commit message
- Removed 'blackbox' related lines
- Removed package.json
- Added tailwind.css
-

**Comment by CMDRyan** [comment]
@Steven-Chan 

Updated README with line: $yarn add express .

**Comment by CMDRyan** [comment]
@Steven-Chan

- Updated README `yarn add express` to `yarn`
- Relocated all source code files to /src and corrected path in app.js
- getJobs.js will now return an array of Job objects

**Pull request merged**